File: problems/140_word_break_ii/Solution.java
import java.util.*;

class Solution {
    private Map<Integer, List<String>> memo = new HashMap<>();
    public List<String> wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        return backtrack(s, wordSet, 0);
    }
    private List<String> backtrack(String s, Set<String> wordSet, int start) {
        if (memo.containsKey(start)) return memo.get(start);
        List<String> result = new ArrayList<>();
        if (start == s.length()) { result.add(""); return result; }
        for (int end = start + 1; end <= s.length(); end++) {
            String word = s.substring(start, end);
            if (wordSet.contains(word)) {
                for (String rest : backtrack(s, wordSet, end)) {
                    result.add(word + (rest.isEmpty() ? "" : " " + rest));
                }
            }
        }
        memo.put(start, result);
        return result;
    }
}

File: problems/140_word_break_ii/solution.py
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        word_set = set(wordDict)
        memo = {}
        def backtrack(start: int) -> List[str]:
            if start in memo:
                return memo[start]
            if start == len(s):
                return ['']
            result = []
            for end in range(start + 1, len(s) + 1):
                word = s[start:end]
                if word in word_set:
                    for rest in backtrack(end):
                        result.append(word + (' ' + rest if rest else ''))
            memo[start] = result
            return result
        return backtrack(0)
